py_zipkin/request_helpers.py

```diff
@@ -0,0 +1,147 @@
+    bits = b3_header.split('-')
+
+    # Handle the lone-sampling-decision case:
+    if len(bits) == 1:
+        parsed['sampled_str'] = bits[0]
+        parsed['sample_decision_only'] = True
+        return parsed
+    if len(bits) > 4:
+        # Too many segments, just return the initialized defaults which will
+        # translate to a None result (no valid incoming span context or
+        # sampling decision)
+        return parsed
+    if len(bits) > 3:
+        parsed['parent_span_id'] = bits[3]
+    if len(bits) > 2:
+        parsed['sampled_str'] = bits[2]
+    if len(bits) > 1:
+        parsed['span_id'] = bits[1]
+        parsed['trace_id'] = bits[0]
+    return parsed
+
+
+def _parse_multi_header(headers):
+    """
+    Parse out and return the data necessary for generating ZipkinAttrs.
+
+    Returns a dict with the following keys:
+        'trace_id':             str or None
+        'span_id':              str or None
+        'parent_span_id':       str or None
+        'sampled_str':          '0', '1', 'd', or None
+        'sample_decision_only': True if header contains only a sampling
+                                decision, false-y value otherwise
+    """
+    parsed = {
+        'trace_id': headers.get('X-B3-TraceId', None),
+        'span_id': headers.get('X-B3-SpanId', None),
+        'parent_span_id': headers.get('X-B3-ParentSpanId', None),
+        'sampled_str': headers.get('X-B3-Sampled', None),
+        'sample_decision_only': False,
+    }
+    if headers.get('X-B3-Flags') == '1':
+        parsed['sampled_str'] = 'd'
+    # Unlike the single-header, "absent" and "present-but-empty-string" are
+    # both representable, and the spec only mentions Absent for encoding
+    # "Defer".  So that means we need to translate an empty-string to a
+    # non-empty, but invalid string.
+    if parsed['sampled_str'] == '':
+        parsed['sampled_str'] = 'INVALID'
+    parsed['sample_decision_only'] = (
+        all(parsed[k] is None for k in ('trace_id', 'span_id', 'parent_span_id'))
+        and parsed['sampled_str'])
+    return parsed
+
+
+def extract_zipkin_attrs_from_headers(headers, sample_rate=100.0,
+                                      use_128bit_trace_id=False):
+    """
+    Implements extraction of B3 headers per:
+        https://github.com/openzipkin/b3-propagation
+
+    The input headers can be any dict-like container that supports "in"
+    membership test and a .get() method that accepts a default value.
+
+    Returns a ZipkinAttrs instance or None
+    """
+    if 'b3' in headers:
+        parsed = _parse_single_header(headers['b3'])
+    else:
+        parsed = _parse_multi_header(headers)
+
+    # Handle the lone-sampling-decision case:
+    if parsed['sample_decision_only']:
+        # See if it's a valid encoding of (only) sample decision
+        if parsed['sampled_str'] in ('0', 'false'):
+            sample_rate = 0.0
+        elif parsed['sampled_str'] in ('1', 'true', 'd'):
```

Comment:
debug wins always, so this moves to the top.. it is also unintuitive to handle false before true..

Reply:
I haven't found that to be the case, myself, but it's easy enough for me to change the order.